api/billingd/filrewards/filrewards.go

```diff
@@ -0,0 +1,217 @@
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(config.DBURI))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to mongo: %v", err)
+	}
+	db := client.Database(config.DBName)
+	col := db.Collection(config.CollectionName)
+	_, err = col.Indexes().CreateMany(ctx, []mongo.IndexModel{
+		{
+			Keys:    bson.D{primitive.E{Key: "key", Value: 1}, primitive.E{Key: "event", Value: 1}},
+			Options: options.Index().SetUnique(true).SetSparse(true),
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "key", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "event", Value: 1}},
+		},
+		{
+			Keys: bson.D{primitive.E{Key: "created_at", Value: 1}},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating indexes: %v", err)
+	}
+	// Populate cache.
+	opts := options.Find()
+	cursor, err := col.Find(ctx, bson.M{}, opts)
+	if err != nil {
+		return nil, fmt.Errorf("querying RewardRecords to populate cache: %s", err)
+	}
+	defer cursor.Close(ctx)
+	cache := make(map[string]map[Event]struct{})
+	for cursor.Next(ctx) {
+		var rec RewardRecord
+		if err := cursor.Decode(&rec); err != nil {
+			return nil, fmt.Errorf("decoding RewardRecord while building cache: %v", err)
+		}
+		ensureKeyEventCache(cache, rec.Key)
+		cache[rec.Key][rec.Event] = struct{}{}
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, fmt.Errorf("iterating cursor while building cache: %v", err)
+	}
+	return &FilRewards{
+		col:          col,
+		rewardsCache: cache,
+	}, nil
+}
+
+func (f *FilRewards) ProcessEvent(ctx context.Context, key string, accountType mdb.AccountType, event analytics.Event) (*RewardRecord, error) {
+	rewardEvent := Unspecified
+	switch event {
+	case analytics.KeyAccountCreated:
+		rewardEvent = FirstKeyAccountCreated
+	case analytics.KeyUserCreated:
+		rewardEvent = FirstKeyUserCreated
+	case analytics.OrgCreated:
+		rewardEvent = FirstOrgCreated
+	case analytics.BillingSetup:
+		rewardEvent = InitialBillingSetup
+	case analytics.BucketCreated:
+		rewardEvent = FirstBucketCreated
+	case analytics.BucketArchiveCreated:
+		rewardEvent = FirstBucketArchiveCreated
+	case analytics.MailboxCreated:
+		rewardEvent = FirstMailboxCreated
+	case analytics.ThreadDbCreated:
+		rewardEvent = FirstThreadDbCreated
+	}
+	if rewardEvent == Unspecified {
+		return nil, nil
+	}
+
+	ensureKeyEventCache(f.rewardsCache, key)
+
+	if _, exists := f.rewardsCache[key][rewardEvent]; exists {
+		// This reward is already granted so bail.
+		return nil, nil
+	}
+
+	rec := RewardRecord{
+		Key:               key,
+		AccountType:       accountType,
+		Event:             rewardEvent,
+		Factor:            eventRewards[rewardEvent].factor,
+		BaseAttoFILReward: baseAttoFILReward,
+		CreatedAt:         time.Now(),
+	}
+
+	if _, err := f.col.InsertOne(ctx, rec); err != nil {
+		return nil, fmt.Errorf("inserting RewardRecord: %v", err)
+	}
+
+	f.rewardsCache[key][rewardEvent] = struct{}{}
+
+	// ToDo: Notify the account owner.
+
+	return &rec, nil
+}
+
```

Comment:
Below are some APIs to query `RewardRecords`. They will be used to build and RPC API for devs to call to view and claim their achievements as well as for APIs for us to do reporting. These APIs can evolve as we understand exactly what information we need and how we need it.

Reply:
these look perfect for the Dashboard, were we can show completed actions checked off etc. 

for the dev, probably just a roll-up like, get total, would be great.